### DISCUSSION TITLE
为何导入已有的评论数据后无法提交新评论

### DISCUSSION THREAD

**Discussion by elkan1788** [walinejs/waline#1148] [Q&A] [answered]
部署环境： Vercel+Leancloud国际版

步骤：

1. 将原来国内版本的评论通过leancloud导出为json格式；
2. 切换到国际版本，导入之前的评论并操作成功；
3. 切换waline评论server地址为新的路径，重新Deploy博客站点；
4. 测试评论提交，出现如下的错误：
```
Invalid value type for field 'insertedAt',value '{"__type":"Date","iso":"2022-06-26T04:42:22.223Z"}',expect type is {:type "String"},but it is '{:type "Date"}'. [400 POST https://lixp20cb.api.lncldglobal.com/1.1/classes/Comment]
```
5. 然后删除导入的评论数据，重新提交评论是可以成功的；
6. 尝试把已有的评论导入到这个新创建的 Class 中显示导入成功，但实际并没有导入；

这个有什么办法处理下吗？

**Comment by lizheming** [comment]
你用的 LeanCloud 的导入导出功能，这块导入导出数据不一致建议找 LeanCloud 提工单咨询。另外直接用 Waline 后台的导入导出功能吧

**Reply by tufu9441** [reply]
> 你用的 LeanCloud 的导入导出功能，这块导入导出数据不一致建议找 LeanCloud 提工单咨询。另外直接用 Waline 后台的导入导出功能吧

以下内容跟此Issue无关，纯属在解决个人历史遗留问题的过程中产生的小发现😂：

我发现在使用CloudBase的数据库时，Waline生成的每一条评论的`_id`是32位的字符串，从CloudBase导出数据再导入LeanCloud时，`_id`作为`objectId`（`String`类型）没有问题（包括我之前Disqus时期评论数据的`_id`是一串10位的纯数字），都可以正常导入和识别；但是，以上数据在导入MongoDB时就会出错，因为MongoDB要求`objectId`是长度为24位的字符串，否则Vercel后台日志中就会出现`Argument passed in must be a single String of 12 bytes or a string of 24 hex characters`的报错，客户端和管理后台也就无法正常展示评论。（直接使用Waline后台的导出和导入功能也会如此，毕竟不会改变数据字段的值。）不过，在使用LeanCloud作为数据库后，产生的新评论的`objectId`就是标准的24位长度的字符串了，这种数据再导入MongoDB的话就不会出错了。

**Reply by lizheming** [reply]
https://github.com/walinejs/waline/blob/main/packages/server/src/controller/db.js#L65-L68 目前针对 SQL 类存储服务，在导入的时候会做 ID 的重置处理，MongoDB 后续可以再优化一下。

**Reply by tufu9441** [reply]
明白了，谢谢！我想着最理想的情况可能是在导出或导入时，无论什么时期和平台的数据，`_id`、`rid`、`pid`这些都能一下子统一成相同的`objectId`格式，不知道是不是想多了😂

**Comment by elkan1788** [comment, marked as answer]
LeanCloud 那边没有办法提供工单，Waline 后台提供的导入方法也是一样会报相同的错误，估计是 API 的问题，另外发现导出的数据会把 objectId, CreateAt 等字段的数据也包括在内，但这些都是保留字段，不能使用的。

```
key[objectId] is reserved
---
Invalid value type for field 'insertedAt',value '"2022-06-14T08:19:54.325Z"',expect type is {:type "Date"},but it is '{:type "String"}'. [400 POST https://lixp20cb.api.lncldglobal.com/1.1/classes/Comment]
```

**Reply by tufu9441** [reply]
> LeanCloud 那边没有办法提供工单，Waline 后台提供的导入方法也是一样会报相同的错误，估计是 API 的问题，另外发现导出的数据会把 objectId, CreateAt 等字段的数据也包括在内，但这些都是保留字段，不能使用的。
> 
> ```
> key[objectId] is reserved
> ---
> Invalid value type for field 'insertedAt',value '"2022-06-14T08:19:54.325Z"',expect type is {:type "Date"},but it is '{:type "String"}'. [400 POST https://lixp20cb.api.lncldglobal.com/1.1/classes/Comment]
> ```

你可以在国内版导出的json文件中尝试把`insertedAt`的字段修改为如下格式，再试着导入一下试试：

`"insertedAt":{"__type":"Date","iso":"2014-03-25T17:28:00.000Z"}`

**Reply by elkan1788** [reply]
谢谢，当时也确实是这样处理的（还好评论不是太多😂），估计是 Leancloud 的 API 有变化影响的。

**Reply by lizheming** [reply]
Waline 后台本身提供导入导出功能，Waline 自身的迁移直接在后台导出，迁移完毕后再进后台导入即可。

**Reply by elkan1788** [reply]
@lizheming 已经明确使用waline导入时，会提示主键不允许写入哦，参考上次回答内容：[discussioncomment-3025026](https://github.com/walinejs/waline/discussions/1148#discussioncomment-3025026)

**Reply by lizheming** [reply]
Waline 的导入需要使用 Waline 的导出数据，而不是 LeanCloud 导出的数据，带有 {__type: "Date"} 这种内容肯定是 LeanCloud 导出的，而不是 Waline 导出的。所以上一个评论重点强调的是操作步骤，而不是只强调了在后台导入。

**Reply by elkan1788** [reply]
😂 前面说的导入导出的数据使用的就是waline后台的功能，并没有混淆导入导出的数据哦。

**Reply by lizheming** [reply]
那你应该就是用错文件了，再 check 下吧，Waline 后台不会导出 `{"__type":"Date","iso":"2022-06-26T04:42:22.223Z"}` 这种字段的，这个很明显是 LeanCloud 后台导出的数据的格式

**Reply by elkan1788** [reply]
你理解错了，{"__type":"Date","iso":"2022-06-26T04:42:22.223Z"} 这个属性是手动添加的，并不是waline导出来的，最后成功的操作也不是通过waline的后台导入导出功能来解决的，就是通过leancloud导出，然后加上前面的字段属性，再用leancloud导入，waline就可以正常显示啦。

**Reply by lizheming** [reply]
好的，后面再看一下
